### PR TITLE
diffcore-rename improvements

### PR DESCRIPTION
**Pull request by newren**
This patch series includes 3 small code cleanups, 1 small bug fix (or perhaps just a UI improvement -- it's a very minor issue either way), more testcases, and 3 performance improvements.  The first 7 patches are relatively easy to review, while the second to last one...is a bit more involved (but provided the biggest performance boost).

Changes since v2:
  * Fixed spelling error: remove unnecessary double c in 'unneccessary' (thanks to Christian for spotting)

cc: Elijah Newren <newren@gmail.com>
cc: Taylor Blau <me@ttaylorr.com>
cc: Christian Couder <christian.couder@gmail.com>

### CODE REVIEW

**Comment by newren** [comment]
That libiconv link breakage affecting vs-build checks for all git PRs is very unfortunate.  But it's clearly not related to my series, so...

File: diffcore-rename.c [hunk 1]
```diff
@@ -447,12 +447,16 @@ static int too_many_rename_candidates(int num_destinations, int num_sources,
...
+	 * the multiplication, the system won't be able to allocate enough
+	 * memory for the matrix anyway.
 	 */
 	if (rename_limit <= 0)
 		rename_limit = 32767;
-	if ((num_destinations <= rename_limit || num_sources <= rename_limit) &&
-	    ((uint64_t)num_destinations * (uint64_t)num_sources
-	     <= (uint64_t)rename_limit * (uint64_t)rename_limit))
+	if (st_mult(num_destinations, num_sources)
+	    <= st_mult(rename_limit, rename_limit))
```
Comment (basarugur) [inline diffcore-rename.c:414]: In cases where the `rename_limit` is already greater than 65535, the `rename_limit * rename_limit` multiplication overflows and process halts. But I don't think an 'overflow error' is very helpful for the users in understanding what is the wrong with their configuration; i.e. `diff.renameLimit` documentation says nothing about a 'maximum allowed value'.

Comment (dscho) [reply diffcore-rename.c:414]: @basarugur please note that the Git project does not do code reviews on GitHub, but on the mailing list instead. You can find the appropriate mail you want to reply to here: https://lore.kernel.org/git/pull.929.v3.git.git.1609272328.gitgitgadget@gmail.com/#r. Then, just follow the guidance at https://github.com/gitgitgadget/gitgitgadget/wiki/ReplyToThis how to send the reply.

**Pull request closed without merging**